rpcs3/rpcs3qt/game_list_frame.cpp

```diff
@@ -580,48 +581,84 @@ void game_list_frame::Refresh(const bool from_drive, const bool scroll_after)
+
 		// Remove duplicates
 		sort(m_path_list.begin(), m_path_list.end());
 		m_path_list.erase(unique(m_path_list.begin(), m_path_list.end()), m_path_list.end());
 
 		const std::string game_icon_path = m_play_hover_movies ? fs::get_config_dir() + "/Icons/game_icons/" : "";
 
-		m_refresh_watcher.setFuture(QtConcurrent::map(m_path_list, [this, cat_unknown_localized = sstr(localized.category.unknown), cat_unknown = sstr(cat::cat_unknown), game_icon_path](const std::string& dir)
+		m_refresh_watcher.setFuture(QtConcurrent::map(m_path_list, [this, dev_flash, cat_unknown_localized = sstr(localized.category.unknown), cat_unknown = sstr(cat::cat_unknown), game_icon_path](const std::string& dir_or_elf)
 		{
+			GameInfo game{};
+			game.path = dir_or_elf;
+
 			const Localized thread_localized;
 
-			const std::string sfo_dir = rpcs3::utils::get_sfo_dir_from_game_path(dir);
+			const std::string sfo_dir = rpcs3::utils::get_sfo_dir_from_game_path(dir_or_elf);
 			const psf::registry psf = psf::load_object(sfo_dir + "/PARAM.SFO");
 			const std::string_view title_id = psf::get_string(psf, "TITLE_ID", "");
 
 			if (title_id.empty())
 			{
-				// Do not care about invalid entries
-				return;
-			}
+				if (!fs::is_file(dir_or_elf))
+				{
+					// Do not care about invalid entries
+					return;
+				}
 
-			GameInfo game{};
-			game.path         = dir;
-			game.serial       = std::string(title_id);
-			game.name         = std::string(psf::get_string(psf, "TITLE", cat_unknown_localized));
-			game.app_ver      = std::string(psf::get_string(psf, "APP_VER", cat_unknown_localized));
-			game.version      = std::string(psf::get_string(psf, "VERSION", cat_unknown_localized));
-			game.category     = std::string(psf::get_string(psf, "CATEGORY", cat_unknown));
-			game.fw           = std::string(psf::get_string(psf, "PS3_SYSTEM_VER", cat_unknown_localized));
-			game.parental_lvl = psf::get_integer(psf, "PARENTAL_LEVEL", 0);
-			game.resolution   = psf::get_integer(psf, "RESOLUTION", 0);
-			game.sound_format = psf::get_integer(psf, "SOUND_FORMAT", 0);
-			game.bootable     = psf::get_integer(psf, "BOOTABLE", 0);
-			game.attr         = psf::get_integer(psf, "ATTRIBUTE", 0);
+				game.serial = dir_or_elf.substr(dir_or_elf.find_last_of(fs::delim) + 1);
+				game.category = cat::cat_ps3_os.toStdString(); // Key for operating system executables
+				game.version = utils::get_firmware_version();
+				game.fw = game.version;
+				game.bootable = 1;
+				game.icon_path = dev_flash + "vsh/resource/explore/icon/icon_home.png";
 
-			if (m_show_custom_icons)
+				if (dir_or_elf.starts_with(dev_flash))
+				{
+					std::string path_vfs = dir_or_elf.substr(dev_flash.size());
+
+					if (usz pos = path_vfs.find_first_not_of(fs::delim); pos != umax && pos != 0)
+					{
+						path_vfs = path_vfs.substr(pos);
+					}
+
+					if (Localized().title.titles.contains(path_vfs))
+					{
+						game.name = Localized().title.titles.at(path_vfs).toStdString();
+					}
+				}
```

Comment:
I don't understand why you're adding this complicated logic when this whole segment is tailored for vsh anyway.

Reply:
There are more stuff in dev_flash other than VSH, this is for them as well.